### PR TITLE
Adding TRTIS streaming API

### PR DESCRIPTION
**Pull request by GuanLuo**
- [x] Add StreamInfer API on server side
- [ ] Discuss how the RPC execution would be different from Infer API
- [ ] Add client side support (C++/Python image client, perf client)
- [ ] Add test cases

### CODE REVIEW

File: src/nvrpc/BidirectionalStreamingLifeCycle.h [hunk 1]
```diff
@@ -0,0 +1,366 @@
+// Copyright (c) 2018, NVIDIA CORPORATION. All rights reserved.
```
Comment (deadeyegoodwin) [inline src/nvrpc/BidirectionalStreamingLifeCycle.h]: copyright year

File: src/core/server.cc [hunk 2]
```diff
@@ -218,6 +219,75 @@ class InferContext final
   }
 };
 
+// For now, this context ExecuteRPC in the same way as InferContext, except that
```
Comment (deadeyegoodwin) [inline src/core/server.cc]: If this code is (mostly) the same then should use a common function. How is the bi-directional streaming being used here?

**Comment by deadeyegoodwin** [comment]
@ryanolson Guan is adding streaming grpc infer API and so is modifying nvrpc. Please take a look.

File: src/core/server.cc (hunk 2 above)
Comment (GuanLuo) [reply src/core/server.cc]: It is the same as the ExecuteRPC(), I didn't know if I will need to change the way to acquire servable if we want the requests to be processed by the same servable instance. If no changes are needed, I agree that we can extract it as a common function.

The bidirectional streaming part is in `BidirectionalStreamingLifeCycle.h`

File: src/nvrpc/Interfaces.h [hunk 3]
```diff
@@ -50,9 +50,12 @@ class IContext {
   static IContext* Detag(void* tag) { return static_cast<IContext*>(tag); }
 
  protected:
-  IContext() = default;
+  IContext() : m_MasterContext(this) {}
+  IContext(IContext* master) : m_MasterContext(master) {}
   void* Tag() { return reinterpret_cast<void*>(this); }
 
+  IContext* m_MasterContext;
```
Comment (ryanolson) [inline src/nvrpc/Interfaces.h]: @GuanLuo - can you help me understand the use case for `m_MasterContext`?  I'd like to keep upstream nvRPC up-to-date.

File: src/nvrpc/BidirectionalStreamingLifeCycle.h [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+}
+
+// The following are a set of functions used as function pointers
+// to keep track of the state of the context.
+template <class Request, class Response>
+bool
+BidirectionalStreamingLifeCycle<Request, Response>::StateInitializedDone(
+    bool ok)
+{
+  if (!ok)
```
Comment (ryanolson) [inline src/nvrpc/BidirectionalStreamingLifeCycle.h]: probably should scope this with {}.  my bad as the reference didn't do it either.

File: src/nvrpc/BidirectionalStreamingLifeCycle.h [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+    m_RequestQueue.emplace();
+  }
+  if (m_NextState == &BidirectionalStreamingLifeCycle<
+                         RequestType, ResponseType>::StateRequestDone)
+    m_NextState = &BidirectionalStreamingLifeCycle<
+        RequestType, ResponseType>::StateResponseDone;
+  m_ReaderWriter->Read(
+      &m_RequestQueue.back(), m_ReadStateContext.IContext::Tag());
+
+  if (should_execute)
```
Comment (ryanolson) [inline src/nvrpc/BidirectionalStreamingLifeCycle.h]: {}

File: src/nvrpc/BidirectionalStreamingLifeCycle.h [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+  bool should_write = true;
+  {
+    std::unique_lock<std::mutex> lock(m_QueueMutex);
+
+    m_ResponseWriteBackQueue.pop();
+    should_write = !m_ResponseWriteBackQueue.empty();
+  }
+  // Only call Write() if the write back queue is not empty,
+  // the FinishResponse() will call Write() otherwise.
+  if (should_write)
```
Comment (ryanolson) [inline src/nvrpc/BidirectionalStreamingLifeCycle.h]: {}

File: src/nvrpc/Interfaces.h (hunk 3 above)
Comment (GuanLuo) [reply src/nvrpc/Interfaces.h]: The original design of nvRPC uses the address of the context as "tag" for completion queue, and the context maintains its own inner state to keep track of what kind of event to be handled next, the life cycle must follow some pattern. 

But then the context can't handle multiple events at the same time, in the case of bidirectional streaming, we can't distinguish if an event is related to Read or Write. So for streaming context, I added two "StateContext" members (derived from IContext) as a way to provides different tags (their address) for handling different events, one for read and one for write.

m_MasterContext is just pointing to the actual streaming context, such that once we know what type of event it is, we call the corresponding functions in the actual context.

File: src/nvrpc/BidirectionalStreamingLifeCycle.h [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+      should_execute = true;
+      m_Executing = true;
+    }
+
+    // Start reading the next request
+    m_RequestQueue.emplace();
+  }
+  if (m_NextState == &BidirectionalStreamingLifeCycle<
+                         RequestType, ResponseType>::StateRequestDone)
+    m_NextState = &BidirectionalStreamingLifeCycle<
```
Comment (ryanolson) [inline src/nvrpc/BidirectionalStreamingLifeCycle.h]: {}

File: src/nvrpc/Interfaces.h (hunk 3 above)
Comment (ryanolson) [reply src/nvrpc/Interfaces.h]: I think I get it.  

`m_MasterContext` is the streams context; however, you have two internal states for reading and writing that you want to independently manage events.  The tags for each of these `StateContext::Tag()` are different, so it allows you to have those two different tags in the completion queue at the same time, even though, they both point back to the same BiDirectionalStreamingContext object.  

LifeCycles are always a bit a mind-warp until you figured out just what's going one.  Good work.

Comment (GuanLuo) [reply src/nvrpc/Interfaces.h]: Yes, that is exactly what it is doing.

File: src/nvrpc/BidirectionalStreamingLifeCycle.h [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+      if (m_Executing)
+        return true;
+    }
+    return false;
+  }
+
+  // Successfully receive request
+  bool should_execute = false;
+  {
+    std::unique_lock<std::mutex> lock(m_QueueMutex);
```
Comment (ryanolson) [inline src/nvrpc/BidirectionalStreamingLifeCycle.h]: no condition variable, so we should probably be using `std::lock_guard` until we upgraded C++17 and then we can use `std::scoped_lock`

File: src/nvrpc/BidirectionalStreamingLifeCycle.h [hunk 9]
```diff
@@ -0,0 +1,366 @@
...
+  // If write didn't go through, then the call is dead. Start reseting
+  if (!ok) {
+    CancelResponse();
+    return true;
+  }
+
+  // Done writing back one response
+  bool should_write = true;
+  {
+    std::unique_lock<std::mutex> lock(m_QueueMutex);
```
Comment (ryanolson) [inline src/nvrpc/BidirectionalStreamingLifeCycle.h]: std::lock_guard

File: src/nvrpc/BidirectionalStreamingLifeCycle.h [hunk 10]
```diff
@@ -0,0 +1,366 @@
...
+}
+
+template <class Request, class Response>
+void
+BidirectionalStreamingLifeCycle<Request, Response>::FinishResponse()
+{
+  bool should_write = false;
+  bool should_execute = true;
+  {
+    std::unique_lock<std::mutex> lock(m_QueueMutex);
```
Comment (ryanolson) [inline src/nvrpc/BidirectionalStreamingLifeCycle.h]: std::lock_guard

File: src/nvrpc/Interfaces.h (hunk 3 above)
Comment (ryanolson) [reply src/nvrpc/Interfaces.h]: 👍 very nice 

I had previously only using the streaming interface for batching, i.e. collecting inputs, compute, then returning a stream of outputs.

This is a great generalization for the more complicated case of independent BiDirectional Reads/Writes.

When you finalize your changes to the internal version of nvRPC, please upstream it to https://github.com/NVIDIA/tensorrt-laboratory/tree/master/tensorrt-laboratory/nvrpc

I'll build a bidirectional flatbuffers example to work out the kinks with the different Message objects.

**Pull request merged**